File: day6/main.cpp
#include <iostream>
#include <vector>
#include <string>
#include <cstdint>
#include <unordered_set>

bool has_cycle(std::vector<std::string> map, std::pair<int, int> start) {
    std::vector<std::vector<int>> visited_dir(map.size(), std::vector<int>(map[0].size(), -1));
    std::pair<int, int> cur_pos = start;
    int cur_dir = 0;
    std::vector<std::pair<int, int>> directions = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};
    visited_dir[start.first][start.second] = 0;
    while (true) {
        std::pair<int, int> next_pos = {cur_pos.first + directions[cur_dir].first, cur_pos.second + directions[cur_dir].second};
        if (next_pos.first < 0 || next_pos.first >= map.size() || next_pos.second < 0 || next_pos.second >= map[0].size()) {
            return false;
        }
        if (map[next_pos.first][next_pos.second] == '#') {
            cur_dir = (cur_dir + 1) % 4;
        } else {
            if (visited_dir[next_pos.first][next_pos.second] == cur_dir) {
                return true;
            }
            visited_dir[next_pos.first][next_pos.second] = cur_dir;
            cur_pos = next_pos;
        }
    }
}

int main() {
    std::vector<std::string> map;
    std::string line;
    while (std::getline(std::cin, line)) {
        map.push_back(line);
    }
    std::pair<int, int> start;
    for (int i = 0; i < map.size(); i++) {
        for (int j = 0; j < map[i].size(); j++) {
            if (map[i][j] == '^') {
                start = {i, j};
                break;
            }
        }
    }
    std::vector<std::pair<int, int>> directions = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};
    std::vector<std::vector<bool>> visited(map.size(), std::vector<bool>(map[0].size(), false));
    visited[start.first][start.second] = true;
    uint32_t res_a = 1;
    std::pair<int, int> cur_pos = start;
    uint32_t cur_dir = 0;
    while (true) {
        std::pair<int, int> next_pos = {cur_pos.first + directions[cur_dir].first, cur_pos.second + directions[cur_dir].second};
        if (next_pos.first < 0 || next_pos.first >= map.size() || next_pos.second < 0 || next_pos.second >= map[0].size()) {
            break;
        }
        if (map[next_pos.first][next_pos.second] == '#') {
            cur_dir = (cur_dir + 1) % 4;
        } else {
            if (!visited[next_pos.first][next_pos.second]) {
                visited[next_pos.first][next_pos.second] = true;
                res_a++;
            }
            cur_pos = next_pos;
        }
    }
    uint32_t res_b = 0;
    for (int i = 0; i < map.size(); i++) {
        for (int j = 0; j < map[i].size(); j++) {
            if (map[i][j] == '.') {
                map[i][j] = '#';
                if (has_cycle(map, start)) {
                    res_b++;
                }
                map[i][j] = '.';
            }
        }
    }
    std::cout << res_a << std::endl;
    std::cout << res_b << std::endl;
    return 0;
}

File: day6/main.rs
use std::io::{self, BufRead};

fn has_loop(map: &Vec<Vec<char>>, start: (usize, usize)) -> bool {
    let mut cur_pos = start;
    let mut cur_dir = 0;
    let dirs: [(i32, i32); 4] = [(-1, 0), (0, 1), (1, 0), (0, -1)];
    let mut visited: Vec<Vec<i32>> =
        vec![vec![-1; map[0].len()]; map.len()];
    visited[start.0][start.1] = 0;

    loop {
        let next_pos = (
            cur_pos.0 as i32 + dirs[cur_dir].0,
            cur_pos.1 as i32 + dirs[cur_dir].1,
        );

        if next_pos.0 < 0
            || next_pos.0 as usize >= map.len()
            || next_pos.1 < 0
            || next_pos.1 as usize >= map[0].len()
        {
            break;
        }
        if map[next_pos.0 as usize][next_pos.1 as usize] == '#' {
            cur_dir = (cur_dir + 1) % 4;
        } else {
            let next_pos = (next_pos.0 as usize, next_pos.1 as usize);
            if visited[next_pos.0][next_pos.1] == cur_dir as i32 {
                return true;
            }
            visited[next_pos.0][next_pos.1] = cur_dir as i32;
            cur_pos = next_pos;
        }
    }
    return false;
}

fn main() {
    let mut map = Vec::new();

    for line in io::stdin().lock().lines() {
        map.push(line.unwrap().chars().collect::<Vec<char>>());
    }

    let mut start = (0, 0);
    'outer: for (i, val) in map.iter().enumerate() {
        for (j, &c) in val.iter().enumerate() {
            if c == '^' {
                start = (i, j);
                break 'outer;
            }
        }
    }

    let mut res_a = 1u32;
    let mut cur_pos = start;
    let mut cur_dir = 0;
    let dirs: [(i32, i32); 4] = [(-1, 0), (0, 1), (1, 0), (0, -1)];
    let mut visited: Vec<Vec<bool>> =
        vec![vec![false; map[0].len()]; map.len()];
    visited[start.0][start.1] = true;

    loop {
        let next_pos = (
            cur_pos.0 as i32 + dirs[cur_dir].0,
            cur_pos.1 as i32 + dirs[cur_dir].1,
        );

        if next_pos.0 < 0
            || next_pos.0 as usize >= map.len()
            || next_pos.1 < 0
            || next_pos.1 as usize >= map[0].len()
        {
            break;
        }
        let next_pos = (next_pos.0 as usize, next_pos.1 as usize);
        if map[next_pos.0][next_pos.1] == '#' {
            cur_dir = (cur_dir + 1) % 4;
        } else {
            if !visited[next_pos.0][next_pos.1] {
                visited[next_pos.0][next_pos.1] = true;
                res_a += 1;
            }
            cur_pos = next_pos;
        }
    }

    let mut res_b = 0;
    for i in 0..map.len() {
        for j in 0..map[0].len() {
            if map[i][j] == '^' || map[i][j] == '#' {
                continue;
            }
            let original = map[i][j];
            map[i][j] = '#';
            if has_loop(&map, start) {
                res_b += 1;
            }
            map[i][j] = original;
        }
    }

    println!("{}", res_a);
    println!("{}", res_b);
}